### PR TITLE
Add `-Werror=non-virtual-dtor` (reland)

### PR DESCRIPTION
**Pull request by malfet**
This PR relands #80584, but instead of adding suppression in CMakeLists.txt suppresses it directly in `llvm_codegen.cpp` and just for a single header.

In general, it's better to avoid `set_target_properties` pattern for suppressing warnings, as it makes build brittle and hard to debug/understand

Test plan: wait for `ciflow/binaries_wheel` to finish

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -784,6 +784,11 @@ if(NOT MSVC)
   string(APPEND CMAKE_CXX_FLAGS " -Wall")
   string(APPEND CMAKE_CXX_FLAGS " -Wextra")
   string(APPEND CMAKE_CXX_FLAGS " -Werror=return-type")
+  if(NOT USE_CUDNN)
```
Comment (huydhn) [inline CMakeLists.txt:787]: For this flag, I will create a follow-up task to fix the issue upstream in cudnn_frontend where the warning comes from. After that is fixed, I think we can remove this not-so-good check

**Review by huydhn — APPROVED** [review]

File: CMakeLists.txt (hunk 1 above)
Comment (malfet) [reply CMakeLists.txt:787]: But can we perhaps suppress one or two files like that?

Comment (huydhn) [reply CMakeLists.txt:787]: FYI, PR on the 3rd party cudnn_frontend https://github.com/NVIDIA/cudnn-frontend/pull/33. When this is accepted and merged, this check can be removed

**Review by kit1980 — APPROVED** [review]

**Pull request closed without merging**